controllers/database_secret_handler.go

```diff
@@ -0,0 +1,160 @@
+			Name:      dbcrName,
+		}})
+	}
+}
+
+func (e *secretEventHandler) Delete(event.DeleteEvent, workqueue.RateLimitingInterface) {
+	logrus.Error("secretEventHandler.Delete(...) event has been FIRED but NOT implemented!")
+}
+func (e *secretEventHandler) Generic(event.GenericEvent, workqueue.RateLimitingInterface) {
+	logrus.Error("secretEventHandler.Generic(...) event has been FIRED but NOT implemented!")
+}
+func (e *secretEventHandler) Create(evt event.CreateEvent, q workqueue.RateLimitingInterface) {
+	logrus.Error("secretEventHandler.Create(...) event has been FIRED but NOT implemented!")
+}
+
+/* ------ Event Filter Functions ------ */
+
+func getDatabasesBySecret(c client.Client, secret types.NamespacedName) ([]kciv1alpha1.Database, []string, error) {
+	databaseList := kciv1alpha1.DatabaseList{}
+	if err := c.List(context.Background(), &databaseList, &client.ListOptions{Namespace: secret.Namespace}); err != nil {
+		return nil, nil, err
+	}
+	var matchedDbs []kciv1alpha1.Database
+	var matchedDbNames []string
+	for _, database := range databaseList.Items {
+		if database.Spec.SecretName == secret.Name {
+			matchedDbs = append(matchedDbs, database)
+			matchedDbNames = append(matchedDbNames, database.Name)
+		}
+	}
+	return matchedDbs, matchedDbNames, nil
+}
+
+func isWatchedNamespace(watchNamespaces []string, ro runtime.Object) bool {
+	if watchNamespaces[0] == "" { // # it's necessary to set "" to watch cluster wide
+		return true // watch for all namespaces
+	}
+	// define object's namespace
+	objectNamespace := ""
+	database, isDatabase := ro.(*kciv1alpha1.Database)
+	if isDatabase {
+		objectNamespace = database.Namespace
+	} else {
+		secret, isSecret := ro.(*corev1.Secret)
+		if isSecret {
+			objectNamespace = secret.Namespace
+		} else {
+			logrus.Info("unknown object", "object", ro)
+			return false
+		}
+	}
+
+	// check that current namespace is watched by db-operator
+	for _, ns := range watchNamespaces {
+		if ns == objectNamespace {
+			return true
+		}
+	}
+	return false
+}
+
+func isDatabase(ro runtime.Object) bool {
+	_, isDatabase := ro.(*kciv1alpha1.Database)
+	return isDatabase
+}
+
+func isObjectUpdated(e event.UpdateEvent) bool {
+	if e.ObjectOld == nil {
+		logrus.Error(nil, "Update event has no old runtime object to update", "event", e)
+		return false
+	}
+	if e.ObjectNew == nil {
+		logrus.Error(nil, "Update event has no new runtime object for update", "event", e)
+		return false
+	}
+	// if object kind is a Database check that 'metadata.generation' field ('spec' section) has been changed
+	_, isDatabase := e.ObjectNew.(*kciv1alpha1.Database)
+	if isDatabase {
+		return e.ObjectNew.GetGeneration() != e.ObjectOld.GetGeneration()
+	}
+
+	// if object kind is a Secret check that password value has changed
+	secretNew, isSecret := e.ObjectNew.(*corev1.Secret)
+	if isSecret {
+		// only annotated secrets are watched
+		annotations := secretNew.ObjectMeta.GetAnnotations()
+		dbcrName := annotations[DbSecretAnnotation]
```

Comment:
check of second return value if key exist, and not skip.

```go
dbcrName, ok := annotations[DbSecretAnnotation]
if not ok {
  return false // no annotation found
}

Reply:
agree, [changed](https://github.com/kloeckner-i/db-operator/pull/130/commits/cb2203fa1d67d05ce0c1997fb9bf4bd74c3efd70)